### PR TITLE
feat: add associative_scan (parallel prefix sum) #3039

### PR DESCRIPTION
**Pull request by cianmag**
## Summary

Implements a general-purpose `associative_scan` operation on `Tensor`, equivalent to [`jax.lax.associative_scan`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.associative_scan.html).

**Algorithm:** Hillis-Steele parallel prefix — O(n log n) work, O(log n) depth. Fully parallelizable.

## API

```python
Tensor.associative_scan(fn, axis=0, reverse=False)
```

- `fn`: Binary associative function `(Tensor, Tensor) -> Tensor`
- `axis`: Scan dimension (default 0)
- `reverse`: Right-to-left scan (default False)

## Examples

```python
# Cumulative sum
Tensor([1, 2, 3, 4]).associative_scan(lambda a, b: a + b)
# → [1, 3, 6, 10]

# Cumulative product
Tensor([1, 2, 3, 4]).associative_scan(lambda a, b: a * b)
# → [1, 2, 6, 24]

# Cumulative max
Tensor([3, 1, 4, 1, 5]).associative_scan(lambda a, b: a.maximum(b))
# → [3, 3, 4, 4, 5]

# Reverse scan
Tensor([1, 2, 3, 4]).associative_scan(lambda a, b: a + b, reverse=True)
# → [10, 9, 7, 4]
```

## Test Results

All tests pass:
- ✅ Addition 1D
- ✅ Multiplication 1D
- ✅ Reverse scan
- ✅ Max scan
- ✅ 2D axis 0
- ✅ 2D axis 1
- ✅ Single element
- ✅ Two elements
- ✅ Power-of-2 lengths (4, 8)
- ✅ Non-power-of-2 lengths (5, 6, 7)
- ✅ Min scan
- ✅ Reverse 2D

## Use Cases

- **Mamba/SSMs**: Parallel scan for state space models
- **CRFs**: Forward-backward algorithm parallelization
- **Kalman filtering**: Parallel smoothing
- **Prefix sums**: General parallel reduction

Closes #3039

### CODE REVIEW

**Comment by geohot** [comment]
NO AI ALLOWED
